### PR TITLE
LLMFineTurningQuestions--NotToMerge

### PR DESCRIPTION
**Pull request by skanjila**
# Code Pull Requests

Please provide the following:

- Some questions from me to start the fine tuning llm workflow discussions, look for ###SXK statements to start our discussions

# Documentation Pull Requests

Note that the documentation HTML files are in `docs/` while the Markdown sources are in `mkdocs/docs`.

If you are proposing a modification to the documentation you should change only the Markdown files.

`api.md` is automatically generated from the docstrings in the code, so if you want to change something in that file, first modify `ludwig/api.py` docstring, then run `mkdocs/code_docs_autogen.py`, which will create `mkdocs/docs/api.md` .

### CODE REVIEW

File: ludwig/decoders/llm_decoders.py [hunk 1]
```diff
@@ -125,6 +125,8 @@ def forward(self, inputs: List[torch.Tensor], **kwargs):
         }
 
 
+###SXK where is this being used, for the decoder how does one build a custom one, one idea here might
+###be for the user to get a list of already available decoders before they start to build one
```
Comment (arnavgarg1) [inline ludwig/decoders/llm_decoders.py:129]: This is being automatically set depending on the output feature type for LLM in the schema definition for that particular output feature type. For e.g. https://github.com/ludwig-ai/ludwig/blob/master/ludwig/schema/features/text_feature.py#L175 for LLM text output features.

Yeah, we'll have this in the docs.

File: ludwig/models/llm.py [hunk 2]
```diff
@@ -46,12 +46,16 @@ def __init__(
 
         logger.info("Loading large language model...")
         self.model = AutoModelForCausalLM.from_pretrained(self.config_obj.model_name)
+        ####SXK should we make this configurable ######
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:49]: This is actually a requirement when using the Ray backend. In theory one could directly load it onto GPUs, but the issue here is that Ray isn't aware of how many GPUs are available to it until later, so we load it onto CPU first and then move it onto GPUs in a later function (to_device). 

In the future, we plan to support meta tensor based initialization since this will not actually load the weights onto the CPU, but just the shapes. Then, if we're doing model parallel training, it'll move the tensors onto the right GPU/CPU devices when it needs to, along wit the optimizer and activations.

File: ludwig/models/llm.py [hunk 3]
```diff
@@ -46,12 +46,16 @@ def __init__(
...
         self.curr_device = torch.device("cpu")  # model initially loaded onto cpu
         logger.info("Done.")
 
+        ####SXK why do we need this adapter and PEFT model, sorry about my ignorance here, how many types
+        ####of adapters can there be
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:54]: PEFT is parameter efficient fine-tuning. It basically allows you to plug and play different "adapters" that are used for fine-tuning as opposed to adjusting the weights of the base LLM model. The forward pass goes through both, but the backward pass just goes through the PEFT adapter weights.

There's many - we already support 5, and more will be added in the future.

File: ludwig/models/llm.py [hunk 4]
```diff
@@ -46,12 +46,16 @@ def __init__(
...
         self.initialize_adapter()
 
         # Determines the maximum length of the context (input + output tokens)
+        ####SXK Is there any validation needed to be done here
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:58]: I think this validation should suffice - what do you have in mind?

File: ludwig/models/llm.py [hunk 5]
```diff
@@ -67,6 +71,7 @@ def __init__(
         # Used only for its metadata about the vocabulary
         self.tokenizer = AutoTokenizer.from_pretrained(self.config_obj.model_name, use_fast=False)
 
+        ####SXK what is the significance of the generation config, again just trying to understand high level context
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:74]: It tells the model how to generate tokens - it controls things such as the temperature, the number of new tokens, the minimum number of tokens, the top p / top k tokens to look it, whether to do greedy decoding or multinomial sampling, etc.

File: ludwig/models/llm.py [hunk 6]
```diff
@@ -95,13 +100,16 @@ def __init__(
         _, self.output_feature_decoder = self.output_features.items()[0]
 
         # ================ Combined loss metric ================
+        ####SXK should we have a class to capture all the loss metrics and output to a dashboard during a run
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:103]: We already do that using a set of predefined metric function classes. These are just maintained separately

File: ludwig/models/llm.py [hunk 7]
```diff
@@ -95,13 +100,16 @@ def __init__(
...
         self.eval_loss_metric = torchmetrics.MeanMetric()
         self.eval_additional_losses_metrics = torchmetrics.MeanMetric()
 
         clear_data_cache()
 
     def initialize_adapter(self):
         """If an adapter config is provided, we want to wrap the model with a PEFT model for fine-tuning."""
+        ###SXK do we think we can extend the peft module but taylor it more towards ludwig
+        ###what would that look like or will this work out of the box
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:112]: This works out of the box already since PEFT is a wrapper on top of transformers, and Ludwig is already compatible with transformers. We could adapt this in the future to a more Ludwig friendly design if we decide we should provide implementations for custom adapters that users might want to design, but I think that's a lot of work and 99.9% users won't want to do that at the moment

File: ludwig/models/llm.py [hunk 8]
```diff
@@ -178,6 +186,9 @@ def get_input_ids(
         """Returns the input ids for the text feature input."""
         return inputs[self.config_obj.input_features[0].name].type(torch.int32)
 
+    ###SXK just a thought here, based on the current use cases does it make sense to have a BaseLLM
+    ###class that we can build custom LLM classes, maybe the base LLM class contains a generic implementation
+    ###of the forward method, lets discuss this some more if it makes sense
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:191]: Hmm, we could, but curious where/how you think that might be useful. I don't see why not though

File: ludwig/models/llm.py [hunk 9]
```diff
@@ -237,6 +248,7 @@ def forward(
 
             return outputs
 
+        ####SXK why dont we need the calculation of gradients here
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:251]: We're computing a forward pass using the model, and we don't need to store the activations since in the generation case, we're not trying to update weights or compute gradients. The torch.no_grad() just freezes the pretrained models weights. Otherwise it's going to generate logits etc which we don't actually need. It also effectively treats the model like it's in "eval" mode

File: ludwig/models/llm.py [hunk 10]
```diff
@@ -317,6 +329,7 @@ def train_loss(
         of_train_losses = {}
         for of_name, of_obj in self.output_features.items():
             _targets, _predictions = targets, predictions
+            ###SXK how can we extend this to other OutputFeature types
             if isinstance(of_obj, TextOutputFeature):
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:333]: It already supports Text and category output features, with more support coming shortly. This if block is specially for text features since they need to be handled separately since they have variable length output. Category/Numbers etc will have fixed length output so this provision doesn't need to be made.

File: ludwig/models/llm.py [hunk 11]
```diff
@@ -407,6 +420,8 @@ def load(self, save_path):
             weights_save_path = os.path.join(save_path, MODEL_WEIGHTS_FILE_NAME)
             config = PeftConfig.from_pretrained(weights_save_path)
             config.inference_mode = False
+            ###SXK I noticed we are tightly coupling to hugging face in places, I wonder
+            ###if we want to make this part configurable as well
```
Comment (arnavgarg1) [inline ludwig/models/llm.py:424]: It is true that we are tighly coupling, but I'm not sure it's worth doing that today since. We could make this a registry pattern where we load LLMs different based on their "types", just like Ludwig does today for a lot of its tokenizers. Additionally, most open source large language models are either on huggingface or on github (and then eventually make their way to huggingface anyway), so this being tightly coupled seems fine for now.

In the future, I do think your suggestion of having a base LLM class and this being a subclass of that called HFLLM or something could make sense though!

File: ludwig/trainers/trainer_llm.py [hunk 12]
```diff
@@ -28,6 +28,7 @@
 logger = logging.getLogger(__name__)
 
 
+###SXK do we want to have a spark trainer as well
```
Comment (arnavgarg1) [inline ludwig/trainers/trainer_llm.py:31]: This would be a pretty big Ludwig contribution, since today we only have support for local and Ray backends for all parts of Ludwig. Would be a great contribution if you're interested in adding it!

File: ludwig/trainers/trainer_llm.py [hunk 13]
```diff
@@ -373,6 +374,8 @@ def run_evaluation(
         return False
 
 
+###SXK I'd like to help out in adding more use cases around this fine tuning workflow, lets talk more about this
+###as well,one thing I was wondering about is how tightly coupled are we to the trainer and the corresponding backend
```
Comment (arnavgarg1) [inline ludwig/trainers/trainer_llm.py:378]: As things go in Ludwig, we are flexible with the following:

Backend: local, ray, horovod, deepspeed
Trainer: local and ray (ray, horovod and deepspeed all run through Ray)
